From 83561e37cf652b673928e5beadeb06dbd2522a26 Mon Sep 17 00:00:00 2001
From: Jactry Zeng <jzeng@codeweavers.com>
Date: Fri, 22 Sep 2023 09:13:29 -0500
Subject: [PATCH 25/42] msxml3: Add GB2312 support for builtin character
 converter.

Imported from wine-staging v11.0 patches/msxml3_encode_gb2312/0001-msxml3-Add-GB2312-support-for-builtin-character-conv.patch.
---
 dlls/msxml3/main.c         | 1 +
 dlls/msxml3/tests/domdoc.c | 8 ++++++++
 2 files changed, 9 insertions(+)

diff --git a/dlls/msxml3/main.c b/dlls/msxml3/main.c
index 16e0b38b185..b0711c510f6 100644
--- a/dlls/msxml3/main.c
+++ b/dlls/msxml3/main.c
@@ -321,6 +321,7 @@ static void init_char_encoders(void)
         xmlCharEncodingOutputFunc output;
     } encoder[] =
     {
+        { "gb2312",       gbk_to_utf8,       utf8_to_gbk       },
         { "gbk",          gbk_to_utf8,       utf8_to_gbk       },
         { "iso8859-1",    iso8859_1_to_utf8, utf8_to_iso8859_1 },
         { "windows-1250", win1250_to_utf8,   utf8_to_win1250   },
diff --git a/dlls/msxml3/tests/domdoc.c b/dlls/msxml3/tests/domdoc.c
index 57f7a510c70..5eeec85e15e 100644
--- a/dlls/msxml3/tests/domdoc.c
+++ b/dlls/msxml3/tests/domdoc.c
@@ -798,6 +798,13 @@ static const char win936xml[] =
 DECL_WIN_936
 "<open></open>";
 
+#define DECL_WIN_GB2312 \
+"<?xml version=\"1.0\" encoding=\"GB2312\"?>"
+
+static const char gb2312xml[] =
+DECL_WIN_GB2312
+"<open></open>";
+
 #define DECL_WIN_1252 \
 "<?xml version=\"1.0\" encoding=\"Windows-1252\"?>"
 
@@ -11007,6 +11014,7 @@ static void test_load(void)
         { iso8859_1_xml, S_OK,    VARIANT_TRUE  },
         { win1252xml,    S_OK,    VARIANT_TRUE  },
         { win936xml,     S_FALSE, VARIANT_FALSE },
+        { gb2312xml,     S_OK,    VARIANT_TRUE  },
     };
 
 
-- 
2.56.0

